playbooks/pki.yml

```diff
@@ -0,0 +1,203 @@
+---
+- name: Prepare CA signing infrastructuren
+  hosts: localhost
+  connection: local
+  gather_facts: no
+  vars:
+    actual_pki_dir: "{{ pki_dir | default('../configuration_files/pki') }}"
+    actual_ca_cn: "{{ ca_cn | default('Hyland - Alfresco signing CA') }}"
+  tasks:
+    - name: Check PKCS12 and CA signing key are given a password
+      ansible.builtin.assert:
+        that:
+          - p12_passphrase
+          - secret_ca_passphrase
+        fail_msg: |
+          this playbook do not generate passwords. You must generate then and
+          store them securely, and pass them as varaiables for the plybook to
+          generate private keys, certificates and containers.
+          expected varaibles are p12_passphrase, secret_ca_passphrase and
+          key_password
+        quiet: true
+
+    - name: Prepare PKI directory structure
+      file:
+        path: "{{ actual_pki_dir }}/{{ item.name }}"
+        state: directory
+        mode: "{{ item.mode | default(omit) }}"
+      loop:
+        - name: ca
+        - name: certificates
+        - name: private
+          mode: "0700"
+    - name: Check for provided CA signing certificate
+      ansible.builtin.stat:
+        path: "{{ ca_file }}"
+      vars:
+        ca_file: >-
+          {{
+            lookup('ansible.builtin.first_found',
+             lookup('ansible.builtin.fileglob', actual_pki_dir + '/ca/*.crt', wantlist=True) +
+             lookup('ansible.builtin.fileglob', actual_pki_dir + '/ca/*.pem', wantlist=True),
+            errors='ignore')
+          }}
+      register: ca_cert_stat
+
+    - name: Check for provided CA signing key
+      ansible.builtin.stat:
+        path: "{{ ca_file }}"
+      vars:
+        ca_file: >-
+          {{
+            lookup('ansible.builtin.first_found',
+             [ actual_pki_dir + '/private/ca.key' ] +
+             lookup('ansible.builtin.fileglob', actual_pki_dir + '/ca/*.key', wantlist=True),
+            errors='ignore')
+          }}
+      register: ca_key_stat
+
+    - name: Validate provided CA certificate and key
+      block:
+        - name: Read CA signing certificate
+          community.crypto.x509_certificate_info:
+            path: "{{ ca_cert_stat.stat.path }}"
+            valid_at:
+              tomorrow: +1d
+          register: ca_cert_valid
+
+        - name: Read CA signing key
+          community.crypto.openssl_privatekey_info:
+            path: "{{ ca_key_stat.stat.path }}"
+            passphrase: "{{ secret_ca_passphrase }}"
+          register: ca_key_valid
+
+        - name: Set CA variables
+          when:
+            - ca_cert_valid.valid_at.tomorrow
+            - ca_key_valid.can_parse_key
+          set_fact:
+            ca_cert_path: "{{ ca_cert_stat.stat.path }}"
+            ca_key_path: "{{ ca_key_stat.stat.path }}"
+      rescue:
```

Comment:
mmm, falling back to self-signed when externally provided pki fails doesn't seems too user friendly: if I want to use my CA but I fail to setup certs in the proper paths I endup with a success play but with a result I was not expecting.

Can we have a dedicate variabile that the user can toggle to declare its desired pki? (self-signed or not)
in that way you would trigger a failure if necessary files are not where they are expected to be

Reply:
I prefer not to add more required variables if we can avoid. However I agree we can simply fail in case we find a user's CA and we can't use it.
In case there is no user CA, then we go with our selfsigned one. wdyt?

Reply:
ok so at least the CA should be put in the proper path and that's the trigger to fail if every other step would fail